cmd/loadtest/main.go

```diff
@@ -0,0 +1,159 @@
+	"context"
+	"encoding/json"
+	"flag"
+	"math/rand"
+	"net/http"
+	"os"
+	"proxy/internal/env"
+
+	"sort"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/rs/zerolog"
+	"github.com/rs/zerolog/log"
+)
+
+type requestRow struct {
+	Id       string `json:"inventoryId"`
+	Priority int    `json:"priority"`
+}
+
+type requestBody struct {
+	InventoryIds []requestRow `json:"inventoryIds"`
+}
+
+func generateRandomPayload(maxRows int) ([]byte, error) {
+	numRows := rand.Intn(maxRows) + 1
+	rows := make([]requestRow, numRows)
+	for i := 0; i < numRows; i++ {
+		id := rand.Intn(10000)
+		rows[i] = requestRow{
+			Id:       "id" + strconv.Itoa(id),
+			Priority: rand.Intn(50),
+		}
+	}
+	body := requestBody{InventoryIds: rows}
+	return json.Marshal(body)
+}
+
+func main() {
+
+	env.LoadEnv()
+	if needTest := os.Getenv("NEED_TEST"); needTest != "true" {
+		return
+	}
+	zerolog.TimeFieldFormat = time.RFC3339
+	log.Logger = log.Output(zerolog.ConsoleWriter{Out: os.Stdout})
+
+	url := "http://localhost:8080/api/v1/measure"
+	concurrency := flag.Int("concurrency", 100, "Number of concurrent workers")
+	duration := flag.Duration("duration", 5*time.Second, "Duration of the load test")
+	maxRows := flag.Int("maxRows", 1000, "Maximum number of rows in a payload")
+	flag.Parse()
+
+	if envURL := os.Getenv("TARGET_URL"); envURL != "" {
+		url = envURL
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), *duration)
+	defer cancel()
+
+	log.Info().
+		Str("target", url).
+		Int("concurrency", *concurrency).
+		Dur("duration", *duration).
+		Msg("Starting load test")
+
+	startTime := time.Now()
+	var wg sync.WaitGroup
+	var reqCount int
+	var reqMu sync.Mutex
+
+	latencyChan := make(chan time.Duration, 100000)
+
+	var latencies []time.Duration
+	var latMu sync.Mutex
+	var aggWg sync.WaitGroup
+	aggWg.Add(1)
+	go func() {
+		defer aggWg.Done()
+		for lat := range latencyChan {
+			latMu.Lock()
+			latencies = append(latencies, lat)
+			latMu.Unlock()
+			reqMu.Lock()
+			reqCount++
+			reqMu.Unlock()
+		}
+
+	}()
+
+	for i := 0; i < *concurrency; i++ {
+		wg.Add(1)
+		go func(workerID int) {
+			defer wg.Done()
+			client := &http.Client{}
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				default:
+				}
+
+				payload, err := generateRandomPayload(*maxRows)
+				if err != nil {
+					log.Error().Err(err).Int("worker", workerID).Msg("Error generating payload")
+					continue
+				}
+
+				req, err := http.NewRequest("POST", url, bytes.NewBuffer(payload))
+				if err != nil {
+					log.Error().Err(err).Int("worker", workerID).Msg("Error creating request")
+					continue
+				}
+				req.Header.Set("Content-Type", "application/json")
+				// Передаем контекст в запрос, чтобы он мог быть отменен по истечении времени теста.
+				req = req.WithContext(ctx)
+
+				reqStart := time.Now()
+				resp, err := client.Do(req)
+				latency := time.Since(reqStart)
+				latencyChan <- latency
+
+				if resp != nil && resp.Body != nil {
+					_ = resp.Body.Close()
```

Comment:
this is usually done after checking for error returned (then no need to check for `resp != nil && resp.Body != nil`)